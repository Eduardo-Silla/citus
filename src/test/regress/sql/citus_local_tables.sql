-- This tests file includes tests for citus local tables.

\set VERBOSITY terse

SET citus.next_shard_id TO 1504000;
SET citus.shard_replication_factor TO 1;
SET citus.enable_local_execution TO ON;
SET citus.log_local_commands TO ON;

CREATE SCHEMA citus_local_tables_test_schema;
SET search_path TO citus_local_tables_test_schema;

------------------------------------------
------- citus local table creation -------
------------------------------------------

-- ensure that coordinator is added to pg_dist_node
SET client_min_messages to ERROR;
SELECT 1 FROM master_add_node('localhost', :master_port, groupId => 0);
RESET client_min_messages;

CREATE TABLE citus_local_table_1 (a int);

-- this should work as coordinator is added to pg_dist_node
SELECT create_citus_local_table('citus_local_table_1');

-- try to remove coordinator and observe failure as there exist a citus local table
SELECT 1 FROM master_remove_node('localhost', :master_port);

DROP TABLE citus_local_table_1;

-- this should work
SELECT 1 FROM master_remove_node('localhost', :master_port);

CREATE TABLE citus_local_table_1 (a int primary key);

-- this should fail as coordinator is removed from pg_dist_node
SELECT create_citus_local_table('citus_local_table_1');

-- let coordinator have citus local tables again for next tests
set client_min_messages to ERROR;
SELECT 1 FROM master_add_node('localhost', :master_port, groupId => 0);
RESET client_min_messages;

-- creating citus local table having no data initially would work
SELECT create_citus_local_table('citus_local_table_1');

-- creating citus local table having data in it would also work
CREATE TABLE citus_local_table_2(a int primary key);
INSERT INTO citus_local_table_2 VALUES(1);

SELECT create_citus_local_table('citus_local_table_2');

-- also create indexes on them
CREATE INDEX citus_local_table_1_idx ON citus_local_table_1(a);
CREATE INDEX citus_local_table_2_idx ON citus_local_table_2(a);

-- drop them for next tests
DROP TABLE citus_local_table_1, citus_local_table_2;

-- create indexes before creating the citus local tables

-- .. for an initially empty table
CREATE TABLE citus_local_table_1(a int);
CREATE INDEX citus_local_table_1_idx ON citus_local_table_1(a);
SELECT create_citus_local_table('citus_local_table_1');

-- .. and for another table having data in it before creating citus local table
CREATE TABLE citus_local_table_2(a int);
INSERT INTO citus_local_table_2 VALUES(1);
CREATE INDEX citus_local_table_2_idx ON citus_local_table_2(a);
SELECT create_citus_local_table('citus_local_table_2');

-- TODO: also show that we do not allow creating citus local tables from mx nodes

-- cannot create citus local table from an existing citus table
CREATE TABLE distributed_table (a int);
SELECT create_distributed_table('distributed_table', 'a');

-- this will error out
SELECT create_citus_local_table('distributed_table');

-- partitiond table tests --

CREATE TABLE partitioned_table(a int, b int) PARTITION BY RANGE (a);
CREATE TABLE partitioned_table_1 PARTITION OF partitioned_table FOR VALUES FROM (0) TO (10);
CREATE TABLE partitioned_table_2 PARTITION OF partitioned_table FOR VALUES FROM (10) TO (20);

-- cannot create partitioned citus local tables
SELECT create_citus_local_table('partitioned_table');

-- cannot create citus local table as a partition of a local table
BEGIN;
  CREATE TABLE citus_local_table PARTITION OF partitioned_table FOR VALUES FROM (20) TO (30);

  -- this should fail
  SELECT create_citus_local_table('citus_local_table');
ROLLBACK;

-- cannot create citus local table as a partition of a local table
-- via ALTER TABLE commands as well
BEGIN;
  CREATE TABLE citus_local_table (a int, b int);

  SELECT create_citus_local_table('citus_local_table');

  -- this should fail
  ALTER TABLE partitioned_table ATTACH PARTITION citus_local_table FOR VALUES FROM (20) TO (30);
ROLLBACK;

-- cannot attach citus local table to a partitioned distributed table
BEGIN;
  SELECT create_distributed_table('partitioned_table', 'a');

  CREATE TABLE citus_local_table (a int, b int);
  SELECT create_citus_local_table('citus_local_table');

  -- this should fail
  ALTER TABLE partitioned_table ATTACH PARTITION citus_local_table FOR VALUES FROM (20) TO (30);
ROLLBACK;

-- drop them for next tests
DROP TABLE citus_local_table_1, citus_local_table_2, distributed_table;

-- create test tables
CREATE TABLE citus_local_table_1 (a int primary key);
SELECT create_citus_local_table('citus_local_table_1');

CREATE TABLE citus_local_table_2 (a int primary key);
SELECT create_citus_local_table('citus_local_table_2');

CREATE TABLE local_table (a int primary key);

CREATE TABLE distributed_table (a int primary key);
SELECT create_distributed_table('distributed_table', 'a');

CREATE TABLE reference_table (a int primary key);
SELECT create_reference_table('reference_table');

-- show that colociation of citus local tables are not supported for now

-- between citus local tables
SELECT mark_tables_colocated('citus_local_table_1', ARRAY['citus_local_table_2']);

-- between citus local tables and reference tables
SELECT mark_tables_colocated('citus_local_table_1', ARRAY['reference_table']);
SELECT mark_tables_colocated('reference_table', ARRAY['citus_local_table_1']);

-- between citus local tables and distributed tables
SELECT mark_tables_colocated('citus_local_table_1', ARRAY['distributed_table']);
SELECT mark_tables_colocated('distributed_table', ARRAY['citus_local_table_1']);

-----------------------------------
---- utility command execution ----
-----------------------------------

-- at this point, we want to see which planner is preferred by citus
SET client_min_messages TO DEBUG2;

-- any foreign key between citus local tables and other tables cannot be set for now
-- each should error out (for now meaningless error messages)

-- between citus local tables
ALTER TABLE citus_local_table_1 ADD CONSTRAINT fkey_c_to_c FOREIGN KEY(a) references citus_local_table_2(a);

-- between citus local tables and reference tables
ALTER TABLE citus_local_table_1 ADD CONSTRAINT fkey_c_to_ref FOREIGN KEY(a) references reference_table(a);
ALTER TABLE reference_table ADD CONSTRAINT fkey_ref_to_c FOREIGN KEY(a) references citus_local_table_1(a);

-- between citus local tables and distributed tables
ALTER TABLE citus_local_table_1 ADD CONSTRAINT fkey_c_to_dist FOREIGN KEY(a) references distributed_table(a);
ALTER TABLE distributed_table ADD CONSTRAINT fkey_dist_to_c FOREIGN KEY(a) references citus_local_table_1(a);

-- between citus local tables and local tables
ALTER TABLE citus_local_table_1 ADD CONSTRAINT fkey_c_to_local FOREIGN KEY(a) references local_table(a);
ALTER TABLE local_table ADD CONSTRAINT fkey_local_to_c FOREIGN KEY(a) references citus_local_table_1(a);

-- prevent verbose logs from DROP command
RESET client_min_messages;

-- drop them for next tests
DROP TABLE citus_local_table_1, citus_local_table_2, distributed_table, local_table, reference_table;

SET client_min_messages TO DEBUG2;

-------------------------------------------------
------- SELECT / INSERT / UPDATE / DELETE -------
-------------------------------------------------

CREATE TABLE citus_local_table(a int, b int);
SELECT create_citus_local_table('citus_local_table');

CREATE TABLE citus_local_table_2(a int, b int);
SELECT create_citus_local_table('citus_local_table_2');

CREATE TABLE reference_table(a int, b int);
SELECT create_reference_table('reference_table');

CREATE TABLE distributed_table(a int, b int);
SELECT create_distributed_table('distributed_table', 'a');

CREATE TABLE local_table(a int, b int);

----------------
---- SELECT ----
----------------

-- join between citus local tables and reference tables would success
SELECT count(*) FROM citus_local_table, reference_table WHERE citus_local_table.a = reference_table.a;

-- join between citus local tables and distributed tables would fail
SELECT citus_local_table.a FROM citus_local_table, distributed_table;

-- join between citus local tables and local tables are okey
SELECT citus_local_table.b, local_table.a
FROM citus_local_table, local_table
WHERE citus_local_table.a = local_table.b;

---------------------------
----- INSERT / SELECT -----
---------------------------

-- those would fail as they are inserting into citus local table
-- from distributed table or reference table

INSERT INTO citus_local_table
SELECT * FROM reference_table;

INSERT INTO citus_local_table
SELECT * from distributed_table;

-- below are ok (simple cases)

INSERT INTO citus_local_table VALUES (1, 2);

INSERT INTO citus_local_table
SELECT * from local_table;

INSERT INTO local_table
SELECT * from citus_local_table;

INSERT INTO citus_local_table
SELECT * FROM citus_local_table_2;

-- INSERT SELECT into distributed or reference table from
-- citus local table is okay, see below two

INSERT INTO distributed_table
SELECT * from citus_local_table;

INSERT INTO reference_table
SELECT * from citus_local_table;

INSERT INTO reference_table
SELECT citus_local_table.a, citus_local_table.b
FROM citus_local_table, local_table
WHERE citus_local_table.a > local_table.b;

---------------------------
----- DELETE / UPDATE -----
---------------------------

-- DELETE/UPDATE commands on local tables and citus local tables are all ok

DELETE FROM citus_local_table
USING local_table
WHERE citus_local_table.b = local_table.b;

UPDATE citus_local_table
SET b = 5
FROM local_table
WHERE citus_local_table.a = 3 AND citus_local_table.b = local_table.b;

-- all should fail (either updating citus local table or reference table or distributed table)
-- as they include citus local table along with other citus tables

UPDATE distributed_table
SET b = 6
FROM citus_local_table
WHERE citus_local_table.a = distributed_table.a;

UPDATE reference_table
SET b = 6
FROM citus_local_table
WHERE citus_local_table.a = reference_table.a;

UPDATE citus_local_table
SET b = 6
FROM distributed_table
WHERE citus_local_table.a = distributed_table.a;

UPDATE citus_local_table
SET b = 6
FROM reference_table
WHERE citus_local_table.a = reference_table.a;

DELETE FROM distributed_table
USING citus_local_table
WHERE citus_local_table.a = distributed_table.a;

DELETE FROM citus_local_table
USING distributed_table
WHERE citus_local_table.a = distributed_table.a;

DELETE FROM reference_table
USING citus_local_table
WHERE citus_local_table.a = reference_table.a;

DELETE FROM citus_local_table
USING reference_table
WHERE citus_local_table.a = reference_table.a;

-- even if we use subquery or cte, they would still fail. see below
DELETE FROM citus_local_table
WHERE citus_local_table.a IN (SELECT a FROM distributed_table);

DELETE FROM citus_local_table
WHERE citus_local_table.a IN (SELECT a FROM reference_table);

WITH distributed_table_cte AS (SELECT * FROM distributed_table)
UPDATE citus_local_table
SET b = 6
FROM distributed_table_cte
WHERE citus_local_table.a = distributed_table_cte.a;

WITH reference_table_cte AS (SELECT * FROM reference_table)
UPDATE citus_local_table
SET b = 6
FROM reference_table_cte
WHERE citus_local_table.a = reference_table_cte.a;

-- prevent verbose logs from DROP command
RESET client_min_messages;

-- cleanup at exit
DROP SCHEMA citus_local_tables_test_schema CASCADE;
