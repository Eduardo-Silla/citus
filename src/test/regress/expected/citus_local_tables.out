-- This tests file includes tests for citus local tables.
\set VERBOSITY terse
SET citus.next_shard_id TO 1504000;
SET citus.shard_replication_factor TO 1;
SET citus.enable_local_execution TO ON;
SET citus.log_local_commands TO ON;
CREATE SCHEMA citus_local_tables_test_schema;
SET search_path TO citus_local_tables_test_schema;
---------------------------------------------------------------------
------- citus local table creation -------
---------------------------------------------------------------------
-- ensure that coordinator is added to pg_dist_node
SET client_min_messages to ERROR;
SELECT 1 FROM master_add_node('localhost', :master_port, groupId => 0);
 ?column?
---------------------------------------------------------------------
        1
(1 row)

RESET client_min_messages;
CREATE TABLE citus_local_table_1 (a int);
-- this should work as coordinator is added to pg_dist_node
SELECT create_citus_local_table('citus_local_table_1');
 create_citus_local_table
---------------------------------------------------------------------

(1 row)

-- try to remove coordinator and observe failure as there exist a citus local table
SELECT 1 FROM master_remove_node('localhost', :master_port);
ERROR:  you cannot remove the primary node of a node group which has shard placements
DROP TABLE citus_local_table_1;
NOTICE:  executing the command locally: DROP TABLE IF EXISTS citus_local_tables_test_schema.citus_local_table_1_xxxxx CASCADE
-- this should work
SELECT 1 FROM master_remove_node('localhost', :master_port);
 ?column?
---------------------------------------------------------------------
        1
(1 row)

CREATE TABLE citus_local_table_1 (a int primary key);
-- this should fail as coordinator is removed from pg_dist_node
SELECT create_citus_local_table('citus_local_table_1');
ERROR:  cannot create citus local table "citus_local_table_1", citus local tables can only be created from coordinator node if it is added to pg_dist_node
-- let coordinator have citus local tables again for next tests
set client_min_messages to ERROR;
SELECT 1 FROM master_add_node('localhost', :master_port, groupId => 0);
 ?column?
---------------------------------------------------------------------
        1
(1 row)

RESET client_min_messages;
-- creating citus local table having no data initially would work
SELECT create_citus_local_table('citus_local_table_1');
 create_citus_local_table
---------------------------------------------------------------------

(1 row)

-- creating citus local table having data in it would also work
CREATE TABLE citus_local_table_2(a int primary key);
INSERT INTO citus_local_table_2 VALUES(1);
SELECT create_citus_local_table('citus_local_table_2');
 create_citus_local_table
---------------------------------------------------------------------

(1 row)

-- also create indexes on them
CREATE INDEX citus_local_table_1_idx ON citus_local_table_1(a);
NOTICE:  executing the command locally: CREATE  INDEX   citus_local_table_1_idx_1504001 ON citus_local_tables_test_schema.citus_local_table_1_1504001 USING btree (a )
CREATE INDEX citus_local_table_2_idx ON citus_local_table_2(a);
NOTICE:  executing the command locally: CREATE  INDEX   citus_local_table_2_idx_1504002 ON citus_local_tables_test_schema.citus_local_table_2_1504002 USING btree (a )
-- drop them for next tests
DROP TABLE citus_local_table_1, citus_local_table_2;
NOTICE:  executing the command locally: DROP TABLE IF EXISTS citus_local_tables_test_schema.citus_local_table_2_xxxxx CASCADE
NOTICE:  executing the command locally: DROP TABLE IF EXISTS citus_local_tables_test_schema.citus_local_table_1_xxxxx CASCADE
-- create indexes before creating the citus local tables
-- .. for an initially empty table
CREATE TABLE citus_local_table_1(a int);
CREATE INDEX citus_local_table_1_idx ON citus_local_table_1(a);
SELECT create_citus_local_table('citus_local_table_1');
 create_citus_local_table
---------------------------------------------------------------------

(1 row)

-- .. and for another table having data in it before creating citus local table
CREATE TABLE citus_local_table_2(a int);
INSERT INTO citus_local_table_2 VALUES(1);
CREATE INDEX citus_local_table_2_idx ON citus_local_table_2(a);
SELECT create_citus_local_table('citus_local_table_2');
 create_citus_local_table
---------------------------------------------------------------------

(1 row)

-- TODO: also show that we do not allow creating citus local tables from mx nodes
-- cannot create citus local table from an existing citus table
CREATE TABLE distributed_table (a int);
SELECT create_distributed_table('distributed_table', 'a');
 create_distributed_table
---------------------------------------------------------------------

(1 row)

-- this will error out
SELECT create_citus_local_table('distributed_table');
ERROR:  table "distributed_table" is already distributed
-- partitiond table tests --
CREATE TABLE partitioned_table(a int, b int) PARTITION BY RANGE (a);
CREATE TABLE partitioned_table_1 PARTITION OF partitioned_table FOR VALUES FROM (0) TO (10);
CREATE TABLE partitioned_table_2 PARTITION OF partitioned_table FOR VALUES FROM (10) TO (20);
-- cannot create partitioned citus local tables
SELECT create_citus_local_table('partitioned_table');
ERROR:  citus local tables can not be involved in a partition relationship
-- cannot create citus local table as a partition of a local table
BEGIN;
  CREATE TABLE citus_local_table PARTITION OF partitioned_table FOR VALUES FROM (20) TO (30);
  -- this should fail
  SELECT create_citus_local_table('citus_local_table');
ERROR:  citus local tables can not be involved in a partition relationship
ROLLBACK;
-- cannot create citus local table as a partition of a local table
-- via ALTER TABLE commands as well
BEGIN;
  CREATE TABLE citus_local_table (a int, b int);
  SELECT create_citus_local_table('citus_local_table');
 create_citus_local_table
---------------------------------------------------------------------

(1 row)

  -- this should fail
  ALTER TABLE partitioned_table ATTACH PARTITION citus_local_table FOR VALUES FROM (20) TO (30);
ERROR:  non-distributed tables cannot have distributed partitions
ROLLBACK;
-- cannot attach citus local table to a partitioned distributed table
BEGIN;
  SELECT create_distributed_table('partitioned_table', 'a');
 create_distributed_table
---------------------------------------------------------------------

(1 row)

  CREATE TABLE citus_local_table (a int, b int);
  SELECT create_citus_local_table('citus_local_table');
 create_citus_local_table
---------------------------------------------------------------------

(1 row)

  -- this should fail
  ALTER TABLE partitioned_table ATTACH PARTITION citus_local_table FOR VALUES FROM (20) TO (30);
ERROR:  distributed tables cannot have non-colocated distributed tables as a partition
ROLLBACK;
-- drop them for next tests
DROP TABLE citus_local_table_1, citus_local_table_2, distributed_table;
NOTICE:  executing the command locally: DROP TABLE IF EXISTS citus_local_tables_test_schema.citus_local_table_2_xxxxx CASCADE
NOTICE:  executing the command locally: DROP TABLE IF EXISTS citus_local_tables_test_schema.citus_local_table_1_xxxxx CASCADE
-- create test tables
CREATE TABLE citus_local_table_1 (a int primary key);
SELECT create_citus_local_table('citus_local_table_1');
 create_citus_local_table
---------------------------------------------------------------------

(1 row)

CREATE TABLE citus_local_table_2 (a int primary key);
SELECT create_citus_local_table('citus_local_table_2');
 create_citus_local_table
---------------------------------------------------------------------

(1 row)

CREATE TABLE local_table (a int primary key);
CREATE TABLE distributed_table (a int primary key);
SELECT create_distributed_table('distributed_table', 'a');
 create_distributed_table
---------------------------------------------------------------------

(1 row)

CREATE TABLE reference_table (a int primary key);
SELECT create_reference_table('reference_table');
 create_reference_table
---------------------------------------------------------------------

(1 row)

-- show that colociation of citus local tables are not supported for now
-- between citus local tables
SELECT mark_tables_colocated('citus_local_table_1', ARRAY['citus_local_table_2']);
ERROR:  citus local tables cannot be colocated with other tables
-- between citus local tables and reference tables
SELECT mark_tables_colocated('citus_local_table_1', ARRAY['reference_table']);
ERROR:  citus local tables cannot be colocated with other tables
SELECT mark_tables_colocated('reference_table', ARRAY['citus_local_table_1']);
ERROR:  citus local tables cannot be colocated with other tables
-- between citus local tables and distributed tables
SELECT mark_tables_colocated('citus_local_table_1', ARRAY['distributed_table']);
ERROR:  citus local tables cannot be colocated with other tables
SELECT mark_tables_colocated('distributed_table', ARRAY['citus_local_table_1']);
ERROR:  citus local tables cannot be colocated with other tables
---------------------------------------------------------------------
---- utility command execution ----
---------------------------------------------------------------------
-- at this point, we want to see which planner is preferred by citus
SET client_min_messages TO DEBUG2;
-- any foreign key between citus local tables and other tables cannot be set for now
-- each should error out (for now meaningless error messages)
-- between citus local tables
ALTER TABLE citus_local_table_1 ADD CONSTRAINT fkey_c_to_c FOREIGN KEY(a) references citus_local_table_2(a);
NOTICE:  executing the command locally: SELECT worker_apply_inter_shard_ddl_command (1504023, 'citus_local_tables_test_schema', 1504024, 'citus_local_tables_test_schema', 'ALTER TABLE citus_local_table_1 ADD CONSTRAINT fkey_c_to_c FOREIGN KEY(a) references citus_local_table_2(a);')
-- between citus local tables and reference tables
ALTER TABLE citus_local_table_1 ADD CONSTRAINT fkey_c_to_ref FOREIGN KEY(a) references reference_table(a);
NOTICE:  executing the command locally: SELECT worker_apply_inter_shard_ddl_command (1504023, 'citus_local_tables_test_schema', 1504029, 'citus_local_tables_test_schema', 'ALTER TABLE citus_local_table_1 ADD CONSTRAINT fkey_c_to_ref FOREIGN KEY(a) references reference_table(a);')
ALTER TABLE reference_table ADD CONSTRAINT fkey_ref_to_c FOREIGN KEY(a) references citus_local_table_1(a);
NOTICE:  executing the command locally: SELECT worker_apply_inter_shard_ddl_command (1504029, 'citus_local_tables_test_schema', 1504023, 'citus_local_tables_test_schema', 'ALTER TABLE reference_table ADD CONSTRAINT fkey_ref_to_c FOREIGN KEY(a) references citus_local_table_1(a);')
-- between citus local tables and distributed tables
ALTER TABLE citus_local_table_1 ADD CONSTRAINT fkey_c_to_dist FOREIGN KEY(a) references distributed_table(a);
ERROR:  cannot create foreign key constraint since foreign keys from reference tables to distributed tables are not supported
ALTER TABLE distributed_table ADD CONSTRAINT fkey_dist_to_c FOREIGN KEY(a) references citus_local_table_1(a);
ERROR:  relation "citus_local_tables_test_schema.citus_local_table_1_1504023" does not exist
-- between citus local tables and local tables
ALTER TABLE citus_local_table_1 ADD CONSTRAINT fkey_c_to_local FOREIGN KEY(a) references local_table(a);
ERROR:  cannot create foreign key constraint
ALTER TABLE local_table ADD CONSTRAINT fkey_local_to_c FOREIGN KEY(a) references citus_local_table_1(a);
DEBUG:  validating foreign key constraint "fkey_local_to_c"
DEBUG:  Creating citus local plan
-- prevent verbose logs from DROP command
RESET client_min_messages;
-- drop them for next tests
DROP TABLE citus_local_table_1, citus_local_table_2, distributed_table, local_table, reference_table;
NOTICE:  executing the command locally: DROP TABLE IF EXISTS citus_local_tables_test_schema.reference_table_xxxxx CASCADE
NOTICE:  executing the command locally: DROP TABLE IF EXISTS citus_local_tables_test_schema.citus_local_table_2_xxxxx CASCADE
NOTICE:  executing the command locally: DROP TABLE IF EXISTS citus_local_tables_test_schema.citus_local_table_1_xxxxx CASCADE
SET client_min_messages TO DEBUG2;
---------------------------------------------------------------------
------- SELECT / INSERT / UPDATE / DELETE -------
---------------------------------------------------------------------
CREATE TABLE citus_local_table(a int, b int);
SELECT create_citus_local_table('citus_local_table');
NOTICE:  executing the command locally: SELECT worker_apply_shard_ddl_command (1504030, 'citus_local_tables_test_schema', 'CREATE TABLE citus_local_tables_test_schema.citus_local_table (a integer, b integer)');SELECT worker_apply_shard_ddl_command (1504030, 'citus_local_tables_test_schema', 'ALTER TABLE citus_local_tables_test_schema.citus_local_table OWNER TO postgres')
 create_citus_local_table
---------------------------------------------------------------------

(1 row)

CREATE TABLE citus_local_table_2(a int, b int);
SELECT create_citus_local_table('citus_local_table_2');
NOTICE:  executing the command locally: SELECT worker_apply_shard_ddl_command (1504031, 'citus_local_tables_test_schema', 'CREATE TABLE citus_local_tables_test_schema.citus_local_table_2 (a integer, b integer)');SELECT worker_apply_shard_ddl_command (1504031, 'citus_local_tables_test_schema', 'ALTER TABLE citus_local_tables_test_schema.citus_local_table_2 OWNER TO postgres')
 create_citus_local_table
---------------------------------------------------------------------

(1 row)

CREATE TABLE reference_table(a int, b int);
SELECT create_reference_table('reference_table');
 create_reference_table
---------------------------------------------------------------------

(1 row)

CREATE TABLE distributed_table(a int, b int);
SELECT create_distributed_table('distributed_table', 'a');
 create_distributed_table
---------------------------------------------------------------------

(1 row)

CREATE TABLE local_table(a int, b int);
---------------------------------------------------------------------
---- SELECT ----
---------------------------------------------------------------------
-- join between citus local tables and reference tables would success
SELECT count(*) FROM citus_local_table, reference_table WHERE citus_local_table.a = reference_table.a;
DEBUG:  Creating citus local plan
NOTICE:  executing the command locally: SELECT count(*) AS count FROM ONLY citus_local_tables_test_schema.citus_local_table_1504030, ONLY citus_local_tables_test_schema.reference_table WHERE (citus_local_table_1504030.a OPERATOR(pg_catalog.=) reference_table.a)
DEBUG:  Creating citus local plan
NOTICE:  executing the command locally: SELECT count(*) AS count FROM ONLY citus_local_tables_test_schema.citus_local_table_1504030, ONLY citus_local_tables_test_schema.reference_table_1504032 WHERE (citus_local_table_1504030.a OPERATOR(pg_catalog.=) reference_table_1504032.a)
 count
---------------------------------------------------------------------
     0
(1 row)

-- join between citus local tables and distributed tables would fail
SELECT citus_local_table.a FROM citus_local_table, distributed_table;
ERROR:  cannot plan SELECT queries with citus local tables and distributed tables
-- join between citus local tables and local tables are okey
SELECT citus_local_table.b, local_table.a
FROM citus_local_table, local_table
WHERE citus_local_table.a = local_table.b;
DEBUG:  Creating citus local plan
NOTICE:  executing the command locally: SELECT citus_local_table_1504030.b, local_table.a FROM ONLY citus_local_tables_test_schema.citus_local_table_1504030, ONLY citus_local_tables_test_schema.local_table WHERE (citus_local_table_1504030.a OPERATOR(pg_catalog.=) local_table.b)
 b | a
---------------------------------------------------------------------
(0 rows)

---------------------------------------------------------------------
----- INSERT / SELECT -----
---------------------------------------------------------------------
-- those would fail as they are inserting into citus local table
-- from distributed table or reference table
INSERT INTO citus_local_table
SELECT * FROM reference_table;
ERROR:  cannot plan INSERT .. SELECT queries to citus local tables selecting from reference tables or distributed tables
INSERT INTO citus_local_table
SELECT * from distributed_table;
ERROR:  cannot plan INSERT .. SELECT queries to citus local tables selecting from reference tables or distributed tables
-- below are ok (simple cases)
INSERT INTO citus_local_table VALUES (1, 2);
DEBUG:  Creating citus local plan
NOTICE:  executing the command locally: INSERT INTO citus_local_tables_test_schema.citus_local_table_1504030 (a, b) VALUES (1, 2)
INSERT INTO citus_local_table
SELECT * from local_table;
DEBUG:  Creating citus local plan
NOTICE:  executing the command locally: INSERT INTO citus_local_tables_test_schema.citus_local_table_1504030 (a, b) SELECT a, b FROM citus_local_tables_test_schema.local_table
INSERT INTO local_table
SELECT * from citus_local_table;
DEBUG:  Creating citus local plan
NOTICE:  executing the command locally: INSERT INTO citus_local_tables_test_schema.local_table (a, b) SELECT a, b FROM citus_local_tables_test_schema.citus_local_table_1504030
INSERT INTO citus_local_table
SELECT * FROM citus_local_table_2;
DEBUG:  Creating citus local plan
NOTICE:  executing the command locally: INSERT INTO citus_local_tables_test_schema.citus_local_table_1504030 (a, b) SELECT a, b FROM citus_local_tables_test_schema.citus_local_table_2_1504031
-- INSERT SELECT into distributed or reference table from
-- citus local table is okay, see below two
INSERT INTO distributed_table
SELECT * from citus_local_table;
DEBUG:  Creating citus local plan
NOTICE:  executing the command locally: INSERT INTO citus_local_tables_test_schema.distributed_table (a, b) SELECT a, b FROM citus_local_tables_test_schema.citus_local_table_1504030
DEBUG:  distributed INSERT ... SELECT can only select from distributed tables
DEBUG:  Collecting INSERT ... SELECT results on coordinator
INSERT INTO reference_table
SELECT * from citus_local_table;
DEBUG:  Creating citus local plan
NOTICE:  executing the command locally: INSERT INTO citus_local_tables_test_schema.reference_table (a, b) SELECT a, b FROM citus_local_tables_test_schema.citus_local_table_1504030
DEBUG:  distributed INSERT ... SELECT can only select from distributed tables
DEBUG:  Collecting INSERT ... SELECT results on coordinator
NOTICE:  executing the copy locally for shard xxxxx
INSERT INTO reference_table
SELECT citus_local_table.a, citus_local_table.b
FROM citus_local_table, local_table
WHERE citus_local_table.a > local_table.b;
DEBUG:  Creating citus local plan
NOTICE:  executing the command locally: INSERT INTO citus_local_tables_test_schema.reference_table (a, b) SELECT citus_local_table_1504030.a, citus_local_table_1504030.b FROM citus_local_tables_test_schema.citus_local_table_1504030, citus_local_tables_test_schema.local_table WHERE (citus_local_table_1504030.a OPERATOR(pg_catalog.>) local_table.b)
DEBUG:  distributed INSERT ... SELECT can only select from distributed tables
DEBUG:  Collecting INSERT ... SELECT results on coordinator
---------------------------------------------------------------------
----- DELETE / UPDATE -----
---------------------------------------------------------------------
-- DELETE/UPDATE commands on local tables and citus local tables are all ok
DELETE FROM citus_local_table
USING local_table
WHERE citus_local_table.b = local_table.b;
DEBUG:  Creating citus local plan
NOTICE:  executing the command locally: DELETE FROM ONLY citus_local_tables_test_schema.citus_local_table_1504030 USING ONLY citus_local_tables_test_schema.local_table WHERE (citus_local_table_1504030.b OPERATOR(pg_catalog.=) local_table.b)
UPDATE citus_local_table
SET b = 5
FROM local_table
WHERE citus_local_table.a = 3 AND citus_local_table.b = local_table.b;
DEBUG:  Creating citus local plan
NOTICE:  executing the command locally: UPDATE ONLY citus_local_tables_test_schema.citus_local_table_1504030 SET b = 5 FROM ONLY citus_local_tables_test_schema.local_table WHERE ((citus_local_table_1504030.a OPERATOR(pg_catalog.=) 3) AND (citus_local_table_1504030.b OPERATOR(pg_catalog.=) local_table.b))
-- all should fail (either updating citus local table or reference table or distributed table)
-- as they include citus local table along with other citus tables
UPDATE distributed_table
SET b = 6
FROM citus_local_table
WHERE citus_local_table.a = distributed_table.a;
ERROR:  cannot plan UPDATE/DELETE queries with citus local tables involving reference tables or distributed tables
UPDATE reference_table
SET b = 6
FROM citus_local_table
WHERE citus_local_table.a = reference_table.a;
ERROR:  cannot plan UPDATE/DELETE queries with citus local tables involving reference tables or distributed tables
UPDATE citus_local_table
SET b = 6
FROM distributed_table
WHERE citus_local_table.a = distributed_table.a;
ERROR:  cannot plan UPDATE/DELETE queries with citus local tables involving reference tables or distributed tables
UPDATE citus_local_table
SET b = 6
FROM reference_table
WHERE citus_local_table.a = reference_table.a;
ERROR:  cannot plan UPDATE/DELETE queries with citus local tables involving reference tables or distributed tables
DELETE FROM distributed_table
USING citus_local_table
WHERE citus_local_table.a = distributed_table.a;
ERROR:  cannot plan UPDATE/DELETE queries with citus local tables involving reference tables or distributed tables
DELETE FROM citus_local_table
USING distributed_table
WHERE citus_local_table.a = distributed_table.a;
ERROR:  cannot plan UPDATE/DELETE queries with citus local tables involving reference tables or distributed tables
DELETE FROM reference_table
USING citus_local_table
WHERE citus_local_table.a = reference_table.a;
ERROR:  cannot plan UPDATE/DELETE queries with citus local tables involving reference tables or distributed tables
DELETE FROM citus_local_table
USING reference_table
WHERE citus_local_table.a = reference_table.a;
ERROR:  cannot plan UPDATE/DELETE queries with citus local tables involving reference tables or distributed tables
-- even if we use subquery or cte, they would still fail. see below
DELETE FROM citus_local_table
WHERE citus_local_table.a IN (SELECT a FROM distributed_table);
ERROR:  cannot plan UPDATE/DELETE queries with citus local tables involving reference tables or distributed tables
DELETE FROM citus_local_table
WHERE citus_local_table.a IN (SELECT a FROM reference_table);
ERROR:  cannot plan UPDATE/DELETE queries with citus local tables involving reference tables or distributed tables
WITH distributed_table_cte AS (SELECT * FROM distributed_table)
UPDATE citus_local_table
SET b = 6
FROM distributed_table_cte
WHERE citus_local_table.a = distributed_table_cte.a;
ERROR:  cannot plan UPDATE/DELETE queries with citus local tables involving reference tables or distributed tables
WITH reference_table_cte AS (SELECT * FROM reference_table)
UPDATE citus_local_table
SET b = 6
FROM reference_table_cte
WHERE citus_local_table.a = reference_table_cte.a;
ERROR:  cannot plan UPDATE/DELETE queries with citus local tables involving reference tables or distributed tables
-- prevent verbose logs from DROP command
RESET client_min_messages;
-- cleanup at exit
DROP SCHEMA citus_local_tables_test_schema CASCADE;
NOTICE:  drop cascades to 9 other objects
